nanoPU_sim.py

```diff
@@ -194,7 +208,12 @@ def start(self):
                 if pkt[NDP].flags.PULL:
                     self.log('Received PULL pkt for tx_msg_id {}, pull offset: {}'.format(tx_msg_id, pkt[NDP].pkt_offset))
                     # increase credit
-                    # TODO: this is not how NDP updates credit, but just to get something running ...
+                    # TODO: this is not how NDP updates credit, but just to get
+                    #       something running ...
+                    # TODO: NDP needs to be able to compare the current pull
+                    #       offset and if pkt[NDP].pkt_offset. If the pkt_offset
+                    #       is greater than the current pull_offset, sets the
+                    #       pull_offset as the pkt_offset.
```

Comment:
PRAW seems to be the correct choice. Here is the corresponding part of the p4 snippet I had written previously:
bit<32> cur_pullOffset;
    @atomic {
      pullOffset_pipeline.read(cur_pullOffset, metadata.flow_indx);
      if (hdr.ndp.sequence_number > cur_pullOffset) {
        cur_pullOffset = hdr.ndp.sequence_number;
        pullOffset_pipeline.write(metadata.flow_indx, cur_pullOffset);
      }

    }

If we are to implement NDP as a fixed function, yes updating the credit event parameters would be sufficient, but this would prevent programmers to implement their choice of update rules because the update rule is implemented in fixed function block now.

NACK packets don't decrease the credit. They just don't increase the credit. We are using credit as the pull_offset value in the case of NDP. Then only PULL packets would increase credit by pulling higher sequence numbers. Here I am talking about the credit state of the sender.

The receiver also keeps pull_offset of the message. This value is updated with data packets. See my comment for line 186 of this code file.

Reply:
This is the only way state is exposed to P4 pipelines - as atomic read-modify-write operations that are supported by the hardware. It's not practical to allow a P4 programmer to express an arbitrary read-modify-write operation because they could easily try to write something that cannot be performed atomically by the hardware. We can allow to P4 programmer to select amongst a set of supported operations by specifying an opcode. So yes, it's not fully programmable, only pseudo programmable, but this is the way all stateful P4 programs are written for hardware targets.

So if we are send a msg which is less than RTT bytes (e.g 10 pkts), then the credit at the sender would be initialized to 10, all pkts would be sent immediately. Then let's say there is an incast and 5 of the pkts are dropped. As the chopped pkts arrive at the receiver, the receiver generates NACKs to indicate which pkts to rtx and the receiver also generates PULLs  which contain pullOffset values 11, 12, 13, 14, 15 and as these PULL pkts arrive at the  sender the credit is increased and the dropped pkts are retransmitted. Does that sound right?

Reply:
Unfortunately it doesn't sound completely right. Here is the exact behavior.
The first 10 packets are sent. Packets 3,4,5,6,7 are dropped. The receiver receives packet 1 and generated ACK+PULL. Same for packet 2. So when the senders receives those acks and pulls, the credit will be increased to 12 and 2 new packets (11, 12) will be sent. Then, the receiver receives packet 3(chopped) which triggers a NACK and a PULL. However the pull_offset in the pull packet is not 13 but 12 again. This is the same for chopped packets of 4,5,6,7. This means when the senders receives the nack and pull of packet 7, the credit will still be 12 but there will be 5 pull packets so that the dropped packets can be re-transmitted.

Here you might be asking why are we sending lots of pull packets with the same pull_offset. The answer is we want to make sure there are 10 packets in flight at any given time for this message in a paced fashion.

Reply:
I see, that makes sense. Thank you. So pkts 3 and 4 will be retransmitted when the PULLs arrive from pkts 1 and 2. And pkts 5, 6, and 7 will be retransmitted after PULLs from pkts 3, 4, and 5 (assuming the network does not reorder any pkts). PULL pkts that arrive from subsequent pkts do not trigger any pkt transmissions.

Reply:
Correct.

Reply:
So the number of pkts generated by a PULL pkt arrival is not simply:
```
num_pkts = pull_pkt.offset - cur_credit
```
because in the case described above, PULL pkts can also generate pkts if `pull_pkt.offset == cur_credit`. What is the right way to compute the number of pkts to generate for each PULL pkt?

Reply:
If we can compute exactly which pkts to transmit when we do the credit state update in the `creditEvent` then we can insert that bitmap (encoding which pkts to transmit) into the active_msgs_fifo, which would eliminate the need to read the credit state in the dequeue() operation of the pktization module. This would make things more amenable to a hardware implementation.

Reply:
1) Response to the first comment:
True. A PULL packet does not tell exactly which packets to transmit. It just permits up to a certain offset. Then it depends on the toBtx bitmap. All the packets up to this offset whose bit is set on toBtx are transmittable now. In short the number of packets that are transmittable (probably not transmitted immediately) is 
```
count_set_bits( [ toBtx & (1 < pull_pkt.offset)-1 ] )
```
2) Response to the second comment:
Yes we could add a copy of ```[ toBtx & (1 < pull_pkt.offset)-1 ]``` to the active_msg_fifo to prevent reading it during dequeue, but I am sceptic about this approach because of 2 reasons:
    * this means we will increase the memory complexity/requirements for the fifo
    * if we deploy a different transport protocol that might reduce the credit in the future (ie, NewReno that halves the window) than this implementation would still send the packets that were allowed previously, but not anymore.

Reply:
Interesting. So this means that the sender doesn't actually need to maintain a `credit` state variable at all? Seems we should change the `creditEvent` into a `schedulePktsEvent`. Rather than performing a read-modify-write on the `credit` state, it will instead compute which pkts to transmit and insert that bitmap into the `active_msgs_fifo, along with the `tx_msg_id`. It will also update the `toBtx` state variable.

Regarding your first concern - yes, the fifo will consume more memory, but I don't think msgs will be more than 100 pkts or so, so the additional memory requirements shouldn't be overwhelming. Regarding your second concern - this may also be a problem but if allows us to implement NDP and Homa then I think it is sufficient. The memory bandwidth savings are so attractive that I think it's worth doing.

Reply:
```schedulePktsEvent```  would still need to keep ```credit``` state because the information from the most recent arriving packet does not necessarily reveal the true ```credit``` information. (ie, the incoming PULL packet is an old one that has pull_offset smaller than the previous PULL packet) Then we must read-modify-write a credit state.

Being able to implement NDP and Homa is NOT sufficient for this project because we don't know if they are good enough for nanoPU environment. We should be as flexible as possible to make sure we don't miss any train after we built the ASIC. So please don't change that at least for now.

Reply:
We also need to move the update of the `toBtx` state variable from the `deliveredEvent` into the `schedulePktsEvent` for NACK pkts. We don't want to access that state from both events.

Reply:
No need at the moment per my explanation above.

Reply:
Ok, so then the pkts to transmit are: `toBtx & (1 << credit)-1`?

One high level comment so that you're aware of what we need to do to actually implement this logic in hardware: We need to make sure that every state variable is accessed in no more than 2 event processing threads (including the msg/pkt processing threads). This is because the FPGA only supports dual port BRAM. There are a few state variables that I think need to be accessed by 3 event processing threads, but one of the threads can tolerate stale read/write operations (e.g. event processing for timeouts).

Also, all of this transport processing will be on the FPGA, not the ASIC, so we will still have flexibility if we decide in the future that our initial implementation is insufficient for some important class of transport protocols.

Reply:
yes the pkts to transmit are: ```toBtx & (1 << credit)-1```